wasm/src/language_service.rs

```diff
@@ -1,40 +1,194 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+use std::{future::Future, path::PathBuf, pin::Pin, sync::Arc};
+
 use crate::{diagnostic::VSDiagnostic, serializable_type};
 use js_sys::JsString;
 use qsc::{self};
+use qsc_project::{EntryType, Manifest, ManifestDescriptor};
+use qsls::JSFileEntry;
 use rustc_hash::FxHashMap;
 use serde::{Deserialize, Serialize};
 use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::JsFuture;
 
 #[wasm_bindgen]
 pub struct LanguageService(qsls::LanguageService<'static>);
 
+/// This macro calls an async JS function, awaits it, and then applies a transformer function to it. Ultimately, it returns a function that accepts a String
+/// and returns a JS promise that is represented by a Rust future. (I know. Ouch. My head.)
+macro_rules! call_async_js_fn {
+    ($js_async_fn: ident, $transformer: expr) => {{
```

Comment:
The name and doc comment are throwing me off here. This code doesn't actually _call_ the function, it just wraps it in a new function that calls it (and does other stuff), right? 

I think the correct name for this would be `async_js_fn_into_async_rust_fn`, but that sounds ridiculously long, so maybe just `into_async_rust_fn`?

`transformer` can be called `map_result` . Since that's basically doing. It's mapping the result value to another value.